---
title: "scRNA-seq pre-processing"
author: "Daan J. Kloosterman"
date: "21/02/2021"
output: html_document
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Pre-processing scRNA-seq data

This module contains the code used to integrate and filter the raw single cell expression data for further downstream processing. The raw single cell expression matrices can be downloaded from (...)

# Loading packages/libraries required for the data pre-processing

```{r}
library(Seurat)
```

# Loading all the raw data files from genomics facility

```{r}
setwd("~/surfdrive/YEAR2_Folder/5. Omics Analysis/1. 10x - GBM Mouse Model/4. Final Data Analysis/0.1 Raw data") # Set working directory to folder where downloaded data is located

load("combineddataset_6106.Rdata") # First CD45-/45+ batch
CD45_1 <- combined

load("Combined_GEX_VDJ.Rda") # Second CD45-/CD45+ batch
CD45_2 <- comb

load("GEX_VDJ_6407_6410_Merged.Rda") # Third CD45-/45+ batch
CD45_3 <- merged

load("ink4a_mg-bmdms.Rdata") # MG/BMDM datasets
mg_bmdm_1 <- combined

load("p53_mg-bmdms.Rda") # MG/BMDM datasets
mg_bmdm_2 <- alldata

rm(alldata, combined, comb, merged)
```

# Running the standard Seurat workflow to prepare data for integration

```{r}
## CD45_1
CD45_1 <- subset(x = CD45_1, subset = nFeature_RNA > 1000)
CD45_1 <- NormalizeData(CD45_1, verbose = FALSE)
CD45_1 <- FindVariableFeatures(CD45_1, selection.method = "vst", 
        nfeatures = 2000, verbose = FALSE)
CD45_1 <- ScaleData(CD45_1)
CD45_1@meta.data$tech <- "CD45_1"


## CD45_2
CD45_2 <- subset(x = CD45_2, subset = nFeature_RNA > 1000)
CD45_2 <- NormalizeData(CD45_2, verbose = FALSE)
CD45_2 <- FindVariableFeatures(CD45_2, selection.method = "vst", 
        nfeatures = 2000, verbose = FALSE)
CD45_2 <- ScaleData(CD45_2)
CD45_2@meta.data$tech <- "CD45_2"

## CD45_3
CD45_3 <- subset(x = CD45_3, subset = nFeature_RNA > 1000)
CD45_3 <- NormalizeData(CD45_3, verbose = FALSE)
CD45_3 <- FindVariableFeatures(CD45_3, selection.method = "vst", 
        nfeatures = 2000, verbose = FALSE)
CD45_3 <- ScaleData(CD45_3)
CD45_3@meta.data$tech <- "CD45_3"

## mg_bmdm_1
mg_bmdm_1 <- subset(x = mg_bmdm_1, subset = nFeature_RNA > 1000)
mg_bmdm_1 <- NormalizeData(mg_bmdm_1, verbose = FALSE)
mg_bmdm_1 <- FindVariableFeatures(mg_bmdm_1, selection.method = "vst", 
        nfeatures = 2000, verbose = FALSE)
mg_bmdm_1 <- ScaleData(mg_bmdm_1)
mg_bmdm_1@meta.data$tech <- "mg_bmdm_1"

## mg_bmdm_2
mg_bmdm_2 <- subset(x = mg_bmdm_2, subset = nFeature_RNA > 1000)
mg_bmdm_2 <- NormalizeData(mg_bmdm_2, verbose = FALSE)
mg_bmdm_2 <- FindVariableFeatures(mg_bmdm_2, selection.method = "vst", 
        nfeatures = 2000, verbose = FALSE)
mg_bmdm_2 <- ScaleData(mg_bmdm_2)
mg_bmdm_2@meta.data$tech <- "mg_bmdm_2"
```

## Data is integrated using integration anchors to correct for batch effect between samples

```{r}
gbm.list <- c(CD45_1, CD45_2, CD45_3, mg_bmdm_1, mg_bmdm_2)
names(gbm.list) <- c("CD45_1", "CD45_2", "CD45_3", "mg_bmdm_1", "mg_bmdm_2")


rm(CD45_1, CD45_2, CD45_3, mg_bmdm_1, mg_bmdm_2)

features <- SelectIntegrationFeatures(object.list = gbm.list, nfeatures = 3000)
reference_dataset <- which(names(gbm.list) == "CD45_3")
gbm.anchors <- FindIntegrationAnchors(object.list = gbm.list,
    anchor.features = features, reference = reference_dataset)

rm(gbm.list)
rm(reference_dataset)

gbm.combined <- IntegrateData(anchorset = gbm.anchors)

rm(gbm.anchors)
```

## Integrated data is filtered based on cell viability (mitochondrial gene expression score) and the nFeature_RNA

```{r}
# Makes a list of mitochondrial genes that start with "Mt-"
mito.genes <- grep(pattern = "Mt-", x = rownames(gbm.myeloid@assays[["RNA"]]), value = TRUE)

# Calculation of mitochondrial gene expression score
percent.mito <- Matrix::colSums(gbm.combined@assays[["RNA"]][mito.genes, ])/Matrix::colSums(gbm.combined@assays[["RNA"]])

# Add mitochondrial score to each cell as a varialbe called "percent.mito"
gbm.combined <- AddMetaData(object = gbm.combined, metadata = percent.mito, col.name = "percent.mito") 

# Visualize cell features, counts and percent.mito per samples to select subset thresholds
Idents(gbm.combined) <- gbm.combined$orig.ident
VlnPlot(object = gbm.combined, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), ncol = 3)
FeatureScatter(object = gbm.combined, feature1 = "nCount_RNA", feature2 = "percent.mito")
FeatureScatter(object = gbm.combined, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

## Design subset based on cutt offs
gbm.combined <- subset(x = gbm.combined, subset = nFeature_RNA > 1000 & nFeature_RNA < 5500 & percent.mito >  -Inf & percent.mito < 0.2 & nCount_RNA < 30000 )

# Downstream analysis on the clustering will be performed on the corrected data, the original unmodified data still resides in the 'RNA' assay and will be used for gene expression comparisons
DefaultAssay(gbm.combined) <- "integrated"
```

## Standard Seurat clustering workflow

```{r}
# Run the standard workflow for visualization and clustering of integrated and filtered single cell datasets
gbm.combined <- ScaleData(gbm.combined, verbose = FALSE)
gbm.combined <- RunPCA(gbm.combined, npcs = 30, verbose = FALSE)
gbm.combined <- FindClusters(gbm.combined, resolution = 0.3, algorithm = 4)
gbm.combined <- RunUMAP(gbm.combined, reduction = "pca", dims = 1:30)
gbm.combined <- FindNeighbors(gbm.combined, reduction = "pca", dims = 1:30)

p1 <- DimPlot(gbm.combined, reduction = "umap", group.by = "seurat_clusters")
p2 <- DimPlot(gbm.combined, reduction = "umap", group.by = "orig.ident", label = TRUE,
    repel = TRUE)
p1 + p2
```

## Adding annotations to each sample

```{r}
## Adding an unique sample name to each sample
gbm.combined$samplename <- plyr::mapvalues(
    x = gbm.combined$orig.ident,  
    from = c("A", "B", "C", "D", "X5765_1_BMDM_P", "X5765_2_MG_P", "X5879_1_BMDM_RC", "X5879_2_MG_RC", "X5879_3_MG_NTB", "S6017", "S6018", "S6035", "6160", "6060", "6190", "6198_1", "6198_2", "6407", "6410_S1", "6410_S2"), 
    to = c("p53_primary_MG", "p53_primary_BMDM", "p53_recurrent_MG", "p53_recurrent_BMDM", "ink4a_primary_BMDM", "ink4a_primary_MG", "ink4a_recurrent_BMDM", "ink4a_recurrent_MG", "ink4a_ntb_MG", "ink4a_recurrent_CD45_a", "ink4a_primary_CD45_a", "ink4a_primary_CD45_b", "ink4a_primary_CD45_c", "p53_primary_CD45_a", "p53_recurrent_CD45", "ink4a_recurrent_CD45_b", "ink4a_recurrent_CD45_c", "p53_primary_CD45_b", "p53_primary_CD45_c", "p53_primary_CD45_d")
) 

## Adding annotation on the stage of the tumor when it was collected
gbm.combined$stage <- plyr::mapvalues(
    x = gbm.combined$orig.ident, 
    from = c("A", "B", "C", "D", "X5765_1_BMDM_P", "X5765_2_MG_P", "X5879_1_BMDM_RC", "X5879_2_MG_RC", "X5879_3_MG_NTB", "S6017", "S6018", "S6035", "6160", "6060", "6190", "6198_1", "6198_2", "6407", "6410_S1", "6410_S2"), 
    to = c("primary", "primary", "recurrent", "recurrent", "primary", "primary", "recurrent", "recurrent", "ntb", "recurrent", "primary", "primary", "primary", "primary", "recurrent", "recurrent", "recurrent", "primary", "primary", "primary")
) 

## Adding annotation on the genetic background of the tumor from which the sample was collected
gbm.combined$model <- plyr::mapvalues(
    x = gbm.combined$orig.ident, 
    from = c("A", "B", "C", "D", "X5765_1_BMDM_P", "X5765_2_MG_P", "X5879_1_BMDM_RC", "X5879_2_MG_RC", "X5879_3_MG_NTB", "S6017", "S6018", "S6035", "6160", "6060", "6190", "6198_1", "6198_2", "6407", "6410_S1", "6410_S2"), 
    to = c("p53", "p53", "p53", "p53", "ink4a", "ink4a", "ink4a", "ink4a", "ink4a", "ink4a", "ink4a", "ink4a", "ink4a", "p53", "p53", "ink4a", "ink4a", "p53", "p53", "p53")
) 
```

## Annotation of unsupervised clusters as tumor cells, astrocytes, endothelial, T-cells, MDMs and MG based on the expression of cell specific markers, respectively Olig2, Gfap, Pecam1, Cd3e, Itga4 and P2ry12. 

```{r}
## Assay is set to "RNA" as we will compare RNA expression between clusters
DefaultAssay(gbm.combined) <- "RNA"
DimPlot(gbm.combined, reduction = "umap",  group.by = "seurat_clusters", label = TRUE)
DotPlot(gbm.combined, group.by = "seurat_clusters", features = c("P2ry12", "Olig1" , "Gfap", "Pecam1", "Cd3e", "Itga4"))

## Give new ID
Idents(gbm.combined) <- gbm.combined$seurat_clusters
new.cluster.ids <- c("MDMs",	"MG",	"Tumor cells",	"Tumor cells",	"MDMs", "MDMs",	"T-cells",	"MG",	"MDMs",	"MG",	"Tumor cells",	"Astrocytes",	"MDMs",  "Tumor cells", "Endothelial cells","MDMs")
names(new.cluster.ids) <- levels(gbm.combined)
gbm.combined <- RenameIdents(gbm.combined, new.cluster.ids)
gbm.combined$celltype <- Idents(gbm.combined) 
my_levels <-  c(	"MG",	"MDMs", "T-cells",	"Endothelial cells", "Astrocytes","Tumor cells")

# Relevel object@ident
Idents(gbm.combined) <- factor(Idents(gbm.combined), levels = my_levels)
gbm.combined$celltype <- Idents(gbm.combined)
DimPlot(gbm.combined, reduction = "umap",  group.by = "celltype", label = TRUE)

## Code to generate table used to calculate the cell type abundance per sample
CellTypeAbundance <- t(table(gbm.combined$celltype, gbm.combined$samplename))
write.csv(CellTypeAbundance, file = "~/CellTypeAbundance.csv")
```


## Split dataset

```{r}
## Dataset is split into a macrophage and tumor cell dataset
gbm.myeloid <- gbm.combined[, gbm.combined$celltype %in% c("MG", "MDMs")]
gbm.tumor  <- gbm.combined[, gbm.combined$celltype %in% c("Tumor cells")]
```


## Saving Data

```{r}
setwd("~/surfdrive/YEAR2_Folder/5. Omics Analysis/1. 10x - GBM Mouse Model/4. Final Data Analysis/0.1 Raw data") # Set working directory to folder where integrated and filtered data will be saved

save(gbm.combined,file="gbm-complete.Rda")
save(gbm.myeloid,file="gbm-complete-myeloid.Rda")
save(gbm.tumor,file="gbm-complete-tumor.Rda")
```
